### PR TITLE
[Feature]Multi cluster warehouse

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 280 / 746 (37.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AlterWarehouseStmt.java|0|15|00.00%|[24, 25, 26, 27, 30, 31, 32, 33, 34, 37, 38, 41, 45, 49, 54]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|0|1|00.00%|[785]|
|:large_blue_circle:|com/starrocks/persist/AlterWhPropertyOplog.java|0|10|00.00%|[32, 35, 36, 37, 38, 41, 45, 50, 51, 54]|
|:large_blue_circle:|com/starrocks/persist/AlterWhClusterOplog.java|0|9|00.00%|[35, 39, 42, 43, 44, 45, 49, 50, 53]|
|:large_blue_circle:|com/starrocks/sql/ast/SuspendWarehouseStmt.java|0|5|00.00%|[20, 21, 22, 25, 30]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|0|6|00.00%|[108, 109, 111, 112, 113, 115]|
|:large_blue_circle:|com/starrocks/warehouse/Cluster.java|0|21|00.00%|[40, 41, 42, 43, 47, 48, 51, 55, 59, 62, 65, 68, 72, 73, 74, 75, 79, 80, 81, 84, 85]|
|:large_blue_circle:|com/starrocks/sql/ast/ResumeWarehouseStmt.java|0|5|00.00%|[20, 21, 22, 25, 30]|
|:large_blue_circle:|com/starrocks/lake/StarOSAgent.java|0|14|00.00%|[401, 403, 404, 405, 406, 407, 408, 409, 410, 412, 414, 452, 457, 461]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|12|00.00%|[201, 202, 203, 208, 209, 210, 214, 215, 216, 219, 220, 221]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|48|08.33%|[152, 153, 154, 155, 158, 159, 160, 161, 162, 165, 166, 167, 168, 169, 172, 173, 174, 175, 176, 179, 181, 182, 183, 184, 187, 189, 190, 191, 192, 195, 196, 197, 198, 199, 1099, 1100, 1103, 1104, 1107, 1108, 1111, 1112, 1115, 1116]|
|:large_blue_circle:|com/starrocks/warehouse/Warehouse.java|59|173|34.10%|[100, 101, 104, 105, 116, 117, 120, 121, 124, 125, 131, 132, 147, 148, 154, 155, 162, 163, 180, 184, 189, 197, 201, 203, 204, 205, 207, 209, 211, 214, 216, 217, 218, 220, 221, 223, 225, 238, 239, 241, 242, 246, 247, 248, 249, 250, 252, 254, 258, 260, 261, 262, 265, 266, 268, 270, 273, 275, 276, 277, 278, 279, 280, 281, 283, 286, 287, 288, 291, 293, 296, 298, 299, 300, 301, 302, 303, 304, 307, 308, 309, 312, 314, 317, 319, 320, 321, 323, 325, 327, 330, 332, 333, 335, 337, 339, 341, 361, 362, 369, 374, 375, 377, 382, 383, 384, 386, 387, 388, 389, 391, 392, 394, 396]|
|:large_blue_circle:|com/starrocks/server/WarehouseManager.java|66|180|36.67%|[92, 133, 134, 136, 138, 141, 143, 144, 145, 147, 149, 153, 154, 155, 156, 157, 158, 159, 160, 163, 164, 165, 166, 167, 170, 171, 172, 173, 174, 175, 176, 177, 180, 181, 182, 183, 184, 203, 205, 206, 207, 209, 211, 214, 216, 217, 219, 222, 223, 225, 228, 229, 230, 232, 235, 236, 237, 239, 242, 243, 244, 245, 246, 247, 249, 250, 253, 254, 255, 258, 259, 260, 265, 266, 267, 270, 271, 274, 275, 276, 277, 280, 281, 284, 285, 288, 289, 293, 294, 295, 299, 301, 302, 303, 304, 305, 306, 308, 310, 311, 312, 313, 314, 315, 318, 319, 327, 328, 329, 335, 340, 341, 343, 355]|
|:large_blue_circle:|com/starrocks/sql/analyzer/WarehouseAnalyzer.java|11|30|36.67%|[30, 37, 38, 44, 48, 49, 56, 57, 58, 60, 65, 66, 67, 69, 73, 74, 75, 77, 84]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|3|8|37.50%|[79, 87, 95, 99, 103]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|4|10|40.00%|[487, 488, 493, 494, 499, 500]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|8|20|40.00%|[832, 833, 834, 835, 840, 841, 842, 843, 848, 849, 850, 851]|
|:large_blue_circle:|com/starrocks/persist/OpWarehouseLog.java|5|11|45.45%|[35, 39, 42, 43, 44, 56]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|20|42|47.62%|[1392, 1393, 1398, 1406, 1407, 1412, 1413, 1414, 1416, 1418, 1420, 1425, 1426, 1427, 1428, 1429, 1430, 1434, 1439, 1440, 1444, 1445]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|19|31|61.29%|[225, 1868, 1870, 1872, 1873, 1875, 1876, 1877, 1880, 1882, 1883, 1884]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|3|66.67%|[1617]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|7|10|70.00%|[1000, 1001, 1002]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|5|7|71.43%|[204, 207]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowWhStmt.java|17|21|80.95%|[40, 41, 42, 50]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateWarehouseStmt.java|13|16|81.25%|[39, 47, 48]|
|:large_blue_circle:|com/starrocks/sql/ast/DropWarehouseStmt.java|10|11|90.91%|[32]|
|:large_blue_circle:|com/starrocks/sql/ast/UseWarehouseStmt.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowClusterStmt.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/ErrorCode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/CaseSensibility.java|1|1|100.00%|[]|

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 496 / 755 (65.70%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|0|1|00.00%|[785]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|0|6|00.00%|[108, 109, 111, 112, 113, 115]|
|:large_blue_circle:|com/starrocks/lake/StarOSAgent.java|0|14|00.00%|[401, 403, 404, 405, 406, 407, 408, 409, 410, 412, 414, 452, 457, 461]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|12|00.00%|[201, 202, 203, 208, 209, 210, 214, 215, 216, 219, 220, 221]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|3|36|08.33%|[223, 1868, 1871, 1872, 1873, 1874, 1875, 1877, 1878, 1879, 1881, 1883, 1884, 1885, 1886, 1887, 1890, 1891, 1892, 1893, 1895, 1896, 1897, 1899, 1901, 1902, 1903, 1904, 1905, 1908, 1909, 1910, 1911]|
|:large_blue_circle:|com/starrocks/warehouse/Cluster.java|6|21|28.57%|[47, 48, 59, 62, 65, 68, 72, 73, 74, 75, 79, 80, 81, 84, 85]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|14|48|29.17%|[152, 153, 154, 155, 158, 159, 160, 161, 162, 165, 166, 167, 168, 169, 172, 173, 174, 175, 176, 179, 181, 182, 183, 184, 187, 189, 190, 191, 192, 195, 196, 197, 198, 199]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|3|8|37.50%|[79, 87, 95, 99, 103]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|8|20|40.00%|[832, 833, 834, 835, 840, 841, 842, 843, 848, 849, 850, 851]|
|:large_blue_circle:|com/starrocks/persist/OpWarehouseLog.java|5|11|45.45%|[35, 39, 42, 43, 44, 56]|
|:large_blue_circle:|com/starrocks/sql/analyzer/WarehouseAnalyzer.java|14|30|46.67%|[30, 37, 38, 44, 48, 49, 58, 65, 66, 67, 69, 73, 74, 75, 77, 84]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|6|50.00%|[643, 1343, 1617]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|6|10|60.00%|[493, 494, 499, 500]|
|:large_blue_circle:|com/starrocks/persist/AlterWhClusterOplog.java|6|9|66.67%|[35, 39, 53]|
|:large_blue_circle:|com/starrocks/persist/AlterWhPropertyOplog.java|7|10|70.00%|[41, 45, 54]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|7|10|70.00%|[1000, 1001, 1002]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|5|7|71.43%|[204, 207]|
|:large_blue_circle:|com/starrocks/warehouse/Warehouse.java|132|174|75.86%|[125, 154, 155, 162, 163, 180, 184, 197, 288, 289, 310, 318, 320, 321, 322, 324, 326, 328, 331, 333, 334, 336, 338, 340, 342, 362, 363, 370, 375, 376, 378, 383, 384, 385, 387, 388, 389, 390, 392, 393, 395, 397]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|32|42|76.19%|[1392, 1393, 1398, 1406, 1407, 1434, 1439, 1440, 1444, 1445]|
|:large_blue_circle:|com/starrocks/sql/ast/SuspendWarehouseStmt.java|4|5|80.00%|[30]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowWhStmt.java|17|21|80.95%|[40, 41, 42, 50]|
|:large_blue_circle:|com/starrocks/sql/ast/ResumeWarehouseStmt.java|4|5|80.00%|[30]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateWarehouseStmt.java|13|16|81.25%|[39, 47, 48]|
|:large_blue_circle:|com/starrocks/sql/ast/AlterWarehouseStmt.java|13|15|86.67%|[37, 38]|
|:large_blue_circle:|com/starrocks/server/WarehouseManager.java|157|180|87.22%|[254, 255, 259, 260, 274, 275, 276, 277, 280, 281, 284, 285, 288, 289, 312, 313, 318, 319, 335, 340, 341, 343, 355]|
|:large_blue_circle:|com/starrocks/sql/ast/DropWarehouseStmt.java|10|11|90.91%|[32]|
|:large_blue_circle:|com/starrocks/sql/ast/UseWarehouseStmt.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowClusterStmt.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/ErrorCode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/CaseSensibility.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 496 / 751 (66.05%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeCheckerV2.java|0|1|00.00%|[785]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|0|6|00.00%|[108, 109, 111, 112, 113, 115]|
|:large_blue_circle:|com/starrocks/lake/StarOSAgent.java|0|14|00.00%|[401, 403, 404, 405, 406, 407, 408, 409, 410, 412, 414, 452, 457, 461]|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|12|00.00%|[201, 202, 203, 208, 209, 210, 214, 215, 216, 219, 220, 221]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|3|36|08.33%|[223, 1868, 1871, 1872, 1873, 1874, 1875, 1877, 1878, 1879, 1881, 1883, 1884, 1885, 1886, 1887, 1890, 1891, 1892, 1893, 1895, 1896, 1897, 1899, 1901, 1902, 1903, 1904, 1905, 1908, 1909, 1910, 1911]|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|14|48|29.17%|[152, 153, 154, 155, 158, 159, 160, 161, 162, 165, 166, 167, 168, 169, 172, 173, 174, 175, 176, 179, 181, 182, 183, 184, 187, 189, 190, 191, 192, 195, 196, 197, 198, 199]|
|:large_blue_circle:|com/starrocks/warehouse/Cluster.java|6|17|35.29%|[59, 62, 72, 73, 74, 75, 79, 80, 81, 84, 85]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|3|8|37.50%|[79, 87, 95, 99, 103]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|8|20|40.00%|[832, 833, 834, 835, 840, 841, 842, 843, 848, 849, 850, 851]|
|:large_blue_circle:|com/starrocks/persist/OpWarehouseLog.java|5|11|45.45%|[35, 39, 42, 43, 44, 56]|
|:large_blue_circle:|com/starrocks/sql/analyzer/WarehouseAnalyzer.java|14|30|46.67%|[30, 37, 38, 44, 48, 49, 58, 65, 66, 67, 69, 73, 74, 75, 77, 84]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|3|6|50.00%|[643, 1343, 1617]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|6|10|60.00%|[493, 494, 499, 500]|
|:large_blue_circle:|com/starrocks/persist/AlterWhClusterOplog.java|6|9|66.67%|[35, 39, 53]|
|:large_blue_circle:|com/starrocks/persist/AlterWhPropertyOplog.java|7|10|70.00%|[41, 45, 54]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|7|10|70.00%|[1000, 1001, 1002]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ShowStmtAnalyzer.java|5|7|71.43%|[204, 207]|
|:large_blue_circle:|com/starrocks/warehouse/Warehouse.java|132|174|75.86%|[125, 154, 155, 162, 163, 180, 184, 197, 288, 289, 310, 318, 320, 321, 322, 324, 326, 328, 331, 333, 334, 336, 338, 340, 342, 362, 363, 370, 375, 376, 378, 383, 384, 385, 387, 388, 389, 390, 392, 393, 395, 397]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|32|42|76.19%|[1392, 1393, 1398, 1406, 1407, 1434, 1439, 1440, 1444, 1445]|
|:large_blue_circle:|com/starrocks/sql/ast/SuspendWarehouseStmt.java|4|5|80.00%|[30]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowWhStmt.java|17|21|80.95%|[40, 41, 42, 50]|
|:large_blue_circle:|com/starrocks/sql/ast/ResumeWarehouseStmt.java|4|5|80.00%|[30]|
|:large_blue_circle:|com/starrocks/sql/ast/CreateWarehouseStmt.java|13|16|81.25%|[39, 47, 48]|
|:large_blue_circle:|com/starrocks/sql/ast/AlterWarehouseStmt.java|13|15|86.67%|[37, 38]|
|:large_blue_circle:|com/starrocks/server/WarehouseManager.java|157|180|87.22%|[254, 255, 259, 260, 274, 275, 276, 277, 280, 281, 284, 285, 288, 289, 312, 313, 318, 319, 335, 340, 341, 343, 355]|
|:large_blue_circle:|com/starrocks/sql/ast/DropWarehouseStmt.java|10|11|90.91%|[32]|
|:large_blue_circle:|com/starrocks/sql/ast/UseWarehouseStmt.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowClusterStmt.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/ErrorCode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectContext.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/CaseSensibility.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    @SerializedName(value = "fullNameToWh")
+    private Map<String, Warehouse> fullNameToWh = new HashMap<>();
+
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+    private final WarehouseProcNode procNode = new WarehouseProcNode();
+    public static final ImmutableList<String> WAREHOUSE_PROC_NODE_TITLE_NAMES = new ImmutableList.Builder<String>()
+            .add("Warehouse")
+            .add("State")
+            .add("Size")
+            .add("min_cluster")
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java:66]: MinCluster

File: fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+            writeUnLock();
+        }
+    }
+
+
+    public void suspendWarehouse(SuspendWarehouseStmt stmt) {
+        String whName = stmt.getFullWhName();
+        readLock();
+        Warehouse warehouse = fullNameToWh.get(whName);
+        readUnlock();
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java:156]: use try finally to ensure unlock operation.

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/DropWarehouseStmt.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.ast;
+
+public class DropWarehouseStmt extends DdlStmt {
+    private final boolean ifNotExists;
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/DropWarehouseStmt.java:18]:
```suggestion
    private final boolean ifExists;
```

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/ShowWhStmt.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import com.starrocks.catalog.ScalarType;
+import com.starrocks.qe.ShowResultSetMetaData;
+
+// Show warehouse statement.
+public class ShowWhStmt extends ShowStmt {
+    private static final String WH_COL = "Warehouse";
+    private static final ShowResultSetMetaData META_DATA =
+            ShowResultSetMetaData.builder()
+                    .addColumn(new Column(WH_COL, ScalarType.createVarchar(256)))
+                    .addColumn(new Column("state", ScalarType.createVarchar(20)))
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/ShowWhStmt.java:29]:
```suggestion
                    .addColumn(new Column("State", ScalarType.createVarchar(20)))
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 5]
```diff
@@ -2146,7 +2193,7 @@ nonReserved
     : AFTER | AGGREGATE | ASYNC | AUTHORS | AVG | ADMIN
     | BACKEND | BACKENDS | BACKUP | BEGIN | BITMAP_UNION | BOOLEAN | BROKER | BUCKETS | BUILTIN
     | CAST | CATALOG | CATALOGS | CEIL | CHAIN | CHARSET | CURRENT | COLLATION | COLUMNS | COMMENT | COMMIT | COMMITTED
-    | COMPUTE | CONNECTION | CONNECTION_ID | CONSISTENT | COSTS | COUNT | CONFIG
+    | COMPUTE | CONNECTION | CONNECTION_ID | CONSISTENT | COSTS | COUNT | CONFIG | CLUSTER | CLUSTERS
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:2196]: better to use alphabetical order.

**Pull request closed without merging**